types/src/protocol_spec.rs

```diff
@@ -0,0 +1,30 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction::{TransactionInfo, TransactionInfoTrait};
+#[cfg(any(test, feature = "fuzzing"))]
+use proptest::arbitrary::Arbitrary;
+#[cfg(any(test, feature = "fuzzing"))]
+use proptest_derive::Arbitrary;
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+
+pub trait ProtocolSpec:
+    Clone + Debug + Serialize + Sync + Send + for<'de> Deserialize<'de>
+{
+    const NAME: &'static str;
+    #[cfg(not(any(test, feature = "fuzzing")))]
+    type TransactionInfo: TransactionInfoTrait;
+    #[cfg(any(test, feature = "fuzzing"))]
+    type TransactionInfo: TransactionInfoTrait + Arbitrary;
```

Comment:
Can we think of a better way to support `Arbitrary`? I think it makes things significantly more confusing if we're conditionally turning requirements on/off.

Reply:
letting those `arb_xx()` functions NOT parameterized on PS addressed this..